modules/core/src/main/java/org/apache/ignite/internal/util/LazyPath.java

```diff
@@ -0,0 +1,145 @@
+        return new LazyPath(() -> defaultPath);
+    }
+
+    /**
+     * Create a new instance from {@code pathSupplier} if it returns a nonempty value, otherwise use {@code defaultPath}.
+     *
+     * @param pathSupplier Path supplier.
+     * @param defaultPath Default path.
+     */
+    public static LazyPath create(Supplier<String> pathSupplier, Path defaultPath) {
+        return new LazyPath(() -> pathOrDefault(pathSupplier.get(), defaultPath));
+    }
+
+    /**
+     * Create a new instance from {@code pathSupplier} if it returns a nonempty value, otherwise use {@code defaultPathSupplier}.
+     *
+     * @param pathSupplier Path supplier.
+     * @param defaultPathSupplier Default path supplier.
+     */
+    public static LazyPath create(Supplier<String> pathSupplier, Supplier<Path> defaultPathSupplier) {
+        return create(pathSupplier, defaultPathSupplier, false);
+    }
+
+    /**
+     * Create a new instance from {@code pathSupplier} if it returns a nonempty value, otherwise use {@code defaultPathSupplier}. If {code
+     * ensureCreated} is {@code true}, then create the path if it doesn't exist.
+     *
+     * @param pathSupplier Path supplier.
+     * @param defaultPathSupplier Default path supplier.
+     * @param ensureCreated If {@code true}, then create the path if it doesn't exist.
+     */
+    public static LazyPath create(Supplier<String> pathSupplier, Supplier<Path> defaultPathSupplier, boolean ensureCreated) {
+        return new LazyPath(() -> {
+            Path path = pathOrDefault(pathSupplier.get(), defaultPathSupplier);
+
+            return ensureCreated ? ensureCreated(path) : path;
+        });
+    }
+
+    /**
+     * Resolve the other path against this one.
+     */
+    public LazyPath resolveLazy(Path other) {
+        return resolveLazy(other, false);
+    }
+
+    /**
+     * Resolve the other path against this one.
+     */
+    public LazyPath resolveLazy(String other) {
+        return resolveLazy(other, false);
+    }
+
+    /**
+     * Resolve the other path against this one.  If {code ensureCreated} is {@code true}, then create the path if it doesn't exist.
+     */
+    public LazyPath resolveLazy(Path other, boolean ensureCreated) {
+        return new LazyPath(() -> {
+            Path resolved = get().resolve(other);
+
+            return ensureCreated ? ensureCreated(resolved) : resolved;
+        });
+    }
+
+    /**
+     * Resolve the other path against this one. If {code ensureCreated} is {@code true}, then create the path if it doesn't exist.
+     */
+    public LazyPath resolveLazy(String other, boolean ensureCreated) {
+        return new LazyPath(() -> {
+            Path resolved = get().resolve(other);
+
+            return ensureCreated ? ensureCreated(resolved) : resolved;
+        });
+    }
+
+    @Override
+    public Path get() {
+        return super.get();
+    }
+
+    private static Path ensureCreated(Path dir) {
```

Comment:
Why do we have two the same methods here and in `org.apache.ignite.internal.configuration.IgnitePaths#ensureCreated`